File: javascript/leetcode/path-sum-ii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */

var pathSum = function(root, sum) {
    paths = []
    var findSum = function(root, thisSum, sum, thisPath) {
        if (root === null) { return; }
        thisSum += root.val;
        thisPath.push(root.val);
        if (root.left === null && root.right === null) {
            if (sum === thisSum) {
                paths.push(thisPath.slice(0));
            }
        } else {
            if (root.left !== null) {
                findSum(root.left, thisSum, sum, thisPath);
            }
            if (root.right !== null) {
                findSum(root.right, thisSum, sum, thisPath);
            }
        }
        thisSum -= root.val;
        thisPath.pop();
    };
    findSum(root, 0, sum, []);
    return paths;
};

File: ruby/leetcode/path-sum-ii.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Integer} sum
# @return {Integer[][]}
def path_sum(root, sum)
    def find_sum(root, this_sum, sum, this_path)
        return if root == nil
        new_sum = this_sum + root.val
        this_path << root.val
        if root.left == nil && root.right == nil
            if sum == new_sum
                @paths << this_path.map(&:clone)
            end
        else
            if root.left != nil
                find_sum(root.left, new_sum, sum, this_path)
            end
            if root.right != nil
                find_sum(root.right, new_sum, sum, this_path)
            end
        end
        this_path.pop()
    end
    @paths = []
    find_sum(root, 0, sum, [])
    return @paths
end
